gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java

```diff
@@ -0,0 +1,239 @@
+/**
+ * A Google cloud storage object.
+ */
+public final class Blob {
+
+  private final Storage storage;
+  private BlobInfo info;
+
+  public static class BlobSourceOption extends Option {
+
+    private static final long serialVersionUID = 214616862061934846L;
+
+    private BlobSourceOption(StorageRpc.Option rpcOption) {
+      super(rpcOption, null);
+    }
+
+    private Storage.BlobSourceOption convert(BlobInfo blobInfo) {
+      switch (rpcOption()) {
+        case IF_GENERATION_MATCH:
+          return Storage.BlobSourceOption.generationMatch(blobInfo.generation());
+        case IF_GENERATION_NOT_MATCH:
+          return Storage.BlobSourceOption.generationNotMatch(blobInfo.generation());
+        case IF_METAGENERATION_MATCH:
+          return Storage.BlobSourceOption.metagenerationMatch(blobInfo.metageneration());
+        case IF_METAGENERATION_NOT_MATCH:
+          return Storage.BlobSourceOption.metagenerationNotMatch(blobInfo.metageneration());
+        default:
+          throw new AssertionError("Unexpected enum value");
+      }
+    }
+
+    public static BlobSourceOption generationMatch() {
+      return new BlobSourceOption(StorageRpc.Option.IF_GENERATION_MATCH);
+    }
+
+    public static BlobSourceOption generationNotMatch() {
+      return new BlobSourceOption(StorageRpc.Option.IF_GENERATION_NOT_MATCH);
+    }
+
+    public static BlobSourceOption metagenerationMatch() {
+      return new BlobSourceOption(StorageRpc.Option.IF_METAGENERATION_MATCH);
+    }
+
+    public static BlobSourceOption metagenerationNotMatch() {
+      return new BlobSourceOption(StorageRpc.Option.IF_METAGENERATION_NOT_MATCH);
+    }
+
+    static Storage.BlobSourceOption[] convert(BlobInfo blobInfo, BlobSourceOption... options) {
+      Storage.BlobSourceOption[] convertedOptions = new Storage.BlobSourceOption[options.length];
+      int index = 0;
+      for (BlobSourceOption option : options) {
+        convertedOptions[index++] = option.convert(blobInfo);
+      }
+      return convertedOptions;
+    }
+  }
+
+  /**
+   * Construct a {@code Blob} object for the provided {@code BlobInfo}. The storage service is used
+   * to issue requests.
+   * 
+   * @param storage the storage service used for issuing requests
+   * @param info blob's info
+   */
+  public Blob(Storage storage, BlobInfo info) {
+    this.storage = checkNotNull(storage);
+    this.info = checkNotNull(info);
+  }
+
+  /**
+   * Construct a {@code Blob} object for the provided bucket and blob names. The storage service is
+   * used to issue requests.
+   * 
+   * @param storage the storage service used for issuing requests
+   * @param bucket bucket's name
+   * @param blob blob's name
+   */
+  public Blob(Storage storage, String bucket, String blob) {
+    this.storage = checkNotNull(storage);
+    this.info = BlobInfo.of(checkNotNull(bucket), checkNotNull(blob));
+  }
+
+  /**
+   * Return the blob's information.
+   */
+  public BlobInfo info() {
+    return info;
+  }
+
```

Comment:
> In these cases adding by default the metagenerationMatch options makes no sense.
Commits up to 2e90926 go the @aozarov way.

I don't think that would be a problem as that info does not have a value for `metageneration`
and therefore will not enforce it. Are you suggesting that it might confuse the caller?
I am actually on the fence on this one.

Options are:
- enforce metageneration check and have only an option to disable this check.
- don't enforce metageneration and keep all options, letting the user choose.

Either way we should be immutable and make the documentation clear.
I want to say that I actually like the first option (as I feel it is slightly more intuitive for blob)
however I also think that it is probably less desired for bucket.

@mziccard pick any option you prefer and lets create a separate issue for it and wait
until we get some more opinions.